core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala

```diff
@@ -672,16 +673,84 @@ class RemoteIndexCacheTest {
     verifyFetchIndexInvocation(count = 1)
   }
 
+  @Test
+  def testConcurrentRemoveReadForCache(): Unit = {
+    // Create a spy Cache Entry
+    val rlsMetadata = new RemoteLogSegmentMetadata(RemoteLogSegmentId.generateNew(idPartition), baseOffset, lastOffset,
+      time.milliseconds(), brokerId, time.milliseconds(), segmentSize, Collections.singletonMap(0, 0L))
+
+    val timeIndex = spy(createTimeIndexForSegmentMetadata(rlsMetadata, new File(tpDir, DIR_NAME)))
+    val txIndex = spy(createTxIndexForSegmentMetadata(rlsMetadata, new File(tpDir, DIR_NAME)))
+    val offsetIndex = spy(createOffsetIndexForSegmentMetadata(rlsMetadata, new File(tpDir, DIR_NAME)))
+
+    val spyEntry = spy(new RemoteIndexCache.Entry(offsetIndex, timeIndex, txIndex))
+    cache.internalCache.put(rlsMetadata.remoteLogSegmentId().id(), spyEntry)
+
+    assertCacheSize(1)
+
+    var entry: RemoteIndexCache.Entry = null
+
+    val latchForCacheRead = new CountDownLatch(1)
+    val latchForCacheRemove = new CountDownLatch(1)
+    val latchForTestWait = new CountDownLatch(1)
+
+    var markForCleanupCallCount = 0
+
+    doAnswer((invocation: InvocationOnMock) => {
+      markForCleanupCallCount += 1
+
+      if (markForCleanupCallCount == 1) {
+        // Signal the CacheRead to unblock itself
+        latchForCacheRead.countDown()
+        // Wait for signal to start renaming the files
+        latchForCacheRemove.await()
+        // Calling the markForCleanup() actual method to start renaming the files
+        invocation.callRealMethod()
```

Comment:
@jeel2420  Why we are invoking this again ` invocation.callRealMethod()` 
It is already called in 
`712 line when(spyEntry).markForCleanup()`

Reply:
invocation.callRealMethod() is called to call the markForCleanup() after read is called and before we start asserting to make sure indexes get renamed before we assert the results.

Reply:
@jeel2420  `markForCleanUp` should be called only one time. We need to test the behaviour when there are concurrent read/remove happens on the cache for the same entry.
In the test we just need to assert the way @showuon  suggested 
// So, maybe we verify with this:
      if (Files.exists(entry.offsetIndex().file().toPath)) {
          assertCacheSize(1)
     } else {
        assertCacheSize(0)
     } 
     Calling  'markForCleanUp' twice asserting CacheSize to be zero which is wrong.

Reply:
@iit2009060 Read `markForCleanUp()` is not getting called twice. Please see the mock, inside that during the first execution of `markForCleanUp()` I am calling the actual markForCleanup() function (i.e Index are getting renamed) but for subsequent calls, mock is doing nothing so we actual markForCleanup() function to rename the indexes is getting called once only and it is as expected. 

I have verified this behaviour as well.

Reply:
@jeel2420 During read `MarkforCleanUp` should not be called not even once as per the functionality.

Then why we need to call it explicitly again  here `invocation.callRealMethod()`. 
I am seeing two invocation of `markForCleanUp`
1. `}).when(spyEntry).markForCleanup()` 712 line no
2. `invocation.callRealMethod()` 708 line no

Reply:
@jeel2420  Is it not called in the 1st line I mentioned 
`}).when(spyEntry).markForCleanup() 712 line no`

Reply:
@iit2009060 I am not quite sure on that.

This mock function is getting called two times. First during the remove call and during that time mock calls the read method to actually rename the index files and it is getting called at the end as well after testWait latch is countdown and during that time mock is doing nothing. 

So I don’t think there is any issue here as actual markForCleanup() method is called only once and indexes are getting renamed before we perform the assertions.

Reply:
@jeel2420  There are two times when markForCleanUp is called. 
1. Throw `remove` function which we are calling in `removeCache Runnable`.
2. One at `invocation.callRealMethod() 708 line no`

Even i tried running your test case locally it always assert with cacheSize 0 , as it is eventually getting deleted. 
IMO we should read and remove concurrently in the separate thread and validate the cacheSize based on the order of execution. We should not need to call explicitly for the scenario. 
@showuon  WDYT ?

Reply:
> There are two times when markForCleanUp is called.
>   1.  remove function which we are calling in removeCache Runnable.
>   2. One at invocation.callRealMethod() 708 line no

You're right, but they are "different" `markForCleanUp`. 
For (1), the `markForCleanUp` is an injected method for controlling the invoking order. 
For (2), it's the real `markForCleanUp` method to rename the cache files.

The goal is to simulate the race condition happened in [KAFKA-15481](https://issues.apache.org/jira/browse/KAFKA-15481).  

> Even i tried running your test case locally it always assert with cacheSize 0 , as it is eventually getting deleted.

Yes, I think so. But in some cases, there could be 1 if getEntry goes after. The thread management are all decided by OS, we can't assure which one will go first, right? 

I think the goal of this test is to make sure the issue in [KAFKA-15481](https://issues.apache.org/jira/browse/KAFKA-15481) will not happen again. That's why I added this [comment](https://github.com/apache/kafka/pull/14483#discussion_r1364844466).

> IMO we should read and remove concurrently in the separate thread and validate the cacheSize based on the order of execution. 

I'm not following you here. What we're doing in this test is to `read and remove concurrently in the separate thread`. About `validate the cacheSize based on the order of execution`, since we can't assure which thread will be executed first, we can't do this, right? If we can decide the execution order, then it means they are not running _concurrently_, is that right?

> We should not need to call explicitly for the scenario.

Maybe you can show us if it were you, what test you'll create. Some pseudo code are enough. Thank you.

Reply:
@showuon  
`For (1), the markForCleanUp is an injected method for controlling the invoking order. So there are latches wait/countdown.`
Do you mean this is a mock method and no rename would happen in this case ?
Effectively the functionality/logic  of `markCleanUp` is called one time only ?

I was thinking something like this 
`val latchForTestWait = new CountDownLatch(2)
 val removeCache = (() => {
      cache.remove(rlsMetadata.remoteLogSegmentId().id())
      latchForTestWait.countdown()
    }): Runnable

    val readCache = (() => {
      entry = cache.getIndexEntry(rlsMetadata)
      // Signal the CacheRemove to start renaming the files
      latchForTestWait.countDown()
    }): Runnable

    val executor = Executors.newFixedThreadPool(2)
    try {
      executor.submit(removeCache: Runnable)
      executor.submit(readCache: Runnable)

      // Wait for signal to complete the test
      latchForTestWait.await()

      // validate cache size `

Reply:
> Do you mean this is a mock method and no rename would happen in this case ?
Effectively the functionality/logic of markCleanUp is called one time only ?

Correct.

> // validate cache size based on the file existence`
>  if offset file exists validate this 
>   // validate rsm call should happen  if( execution order is remove,read)
>   if cache size == 0
>   // validate no rsm call should happen if ( execution order is read,remove)

Yes, they are basically similar with what we have now. By injecting mock implementation for `markForCleanUp` is just to make the 2 thread execution more close. In the end, what we have now is to invoke "realMethod", which is what you did above. I'm fine if you think we should velidate the rsm call count. But again, they are basically testing the same thing.

Reply:
Yes correct , it is  testing the same thing. I am also fine. But from readability perspective the one I propose is simpler to understand. I left @jeel2420  to make a decision here.

Reply:
@iit2009060 As current test case is able to reproduce the case mentioned in the jira [KAFKA-15481](https://issues.apache.org/jira/browse/KAFKA-15481) I think we should be fine. 

The only reason to have `markForCleanUp` mock is to have control over the 2 thread execution.